### ISSUE TITLE
hypothesis文件生成有问题

### ISSUE THREAD

**Issue by lm089381** [lihongzhao99/Context-Enhanced-Framework#2]
您好，很抱歉打扰您！我想问一下复现这个代码，是先用PHASE=‘TRAIN'进行训练，然后将这个参数值改成’INFER‘生成hypothesis和references文件，再使用nlg-eval进行测试吗？
为什么我这样操作之后，生成的hypothesis全部是重复单词，最后的评价结果为0.000

**Comment by lihongzhao99** [comment]
这可能是模型还没有拟合好的原因， 你能发来你的训练log吗？

**Comment by H1ghDefinition** [comment]
您好，我在IU XRay数据集上的复现也同样遇到这个问题。模型训练在前25个epoch的loss都不断下降，第25个epoch的验证loss为1.37并保存了训练权重，后续epoch出现了过拟合的情况，训练loss在最后一个epoch下降至0.84而验证loss升高至1.44，训练log如附件所示。

[log.txt](https://github.com/user-attachments/files/21114502/log.txt)

**Comment by H1ghDefinition** [comment]
> 您好，我在IU XRay数据集上的复现也同样遇到这个问题。模型训练在前25个epoch的loss都不断下降，第25个epoch的验证loss为1.37并保存了训练权重，后续epoch出现了过拟合的情况，训练loss在最后一个epoch下降至0.84而验证loss升高至1.44，训练log如附件所示。
> 
> [log.txt](https://github.com/user-attachments/files/21114502/log.txt)

顺带附上我生成的三个结果文件：

[x_IUXRAY_Context_DenseNet121_MaxView2_NumLabel114_History_Hyp.txt](https://github.com/user-attachments/files/21114517/x_IUXRAY_Context_DenseNet121_MaxView2_NumLabel114_History_Hyp.txt)

[x_IUXRAY_Context_DenseNet121_MaxView2_NumLabel114_History_Lbl.txt](https://github.com/user-attachments/files/21114519/x_IUXRAY_Context_DenseNet121_MaxView2_NumLabel114_History_Lbl.txt)

[x_IUXRAY_Context_DenseNet121_MaxView2_NumLabel114_History_Ref.txt](https://github.com/user-attachments/files/21114522/x_IUXRAY_Context_DenseNet121_MaxView2_NumLabel114_History_Ref.txt)

**Comment by lihongzhao99** [comment]
问题可能出在这部分代码：

    def infer(self, source_embed, source_embed2, source_pad_mask=None, max_len=100, top_k=1, bos_id=1, pad_id=3):
        outputs = torch.ones((top_k, source_embed.shape[0], 1), dtype=torch.long).to(source_embed.device) * bos_id
        scores = torch.zeros((top_k, source_embed.shape[0]), dtype=torch.float32).to(source_embed.device)
        for _ in range(1,max_len):
            possible_outputs = []
            possible_scores = []
            for k in range(top_k):
                output = outputs[k]
                score = scores[k]
                att, emb = self.forward(source_embed, source_embed2, output, source_pad_mask=source_pad_mask, target_pad_mask=(output == pad_id))
                val, idx = torch.topk(att[:,-1,:], top_k)
                log_val = -torch.log(val)
                for i in range(top_k):
                    new_output = torch.cat([output, idx[:,i].view(-1,1)], dim=-1)
                    new_score = score + log_val[:,i].view(-1)
                    possible_outputs.append(new_output.unsqueeze(0))
                    possible_scores.append(new_score.unsqueeze(0))
            possible_outputs = torch.cat(possible_outputs, dim=0)
            possible_scores = torch.cat(possible_scores, dim=0)
            val, idx = torch.topk(possible_scores, top_k, dim=0)
            col_idx = torch.arange(idx.shape[1], device=idx.device).unsqueeze(0).repeat(idx.shape[0],1)
            outputs = possible_outputs[idx,col_idx]
            scores = possible_scores[idx,col_idx]
        val, idx = torch.topk(scores, 1, dim=0)
        col_idx = torch.arange(idx.shape[1], device=idx.device).unsqueeze(0).repeat(idx.shape[0],1)
        output = outputs[idx,col_idx]
        return output.squeeze(0)


我会在空闲时修复。

**Comment by lihongzhao99** [comment]
我初步修复了这个错误，提交了修复后新的文件。
